feature(targetlocker/mysql)!: Add MySQL target locker

xaionaro opened the pull request:
Added a plugin to handle target-locking in an MySQL table.

ITS: https://github.com/facebookincubator/contest/issues/35

Additional notes:
* A flag `-targetLocker` was added to `./cmds/contest/`. By default it uses the same DSN as `-dbURI`.
* Proposed a new way to handle factories. It will allow to check for duplicates and some other handy stuff.

BREAKING CHANGE:
* Interface `target.Locker` was changed.
* `target.LockerFactory` was converted to an interface.

xaionaro commented on pkg/runner/job_runner.go:
```diff
@@ -105,13 +105,20 @@ func (jr *JobRunner) Run(j *job.Job) ([][]*job.Report, []*job.Report, error) {
+				if err != nil {
+					errCh <- fmt.Errorf("unable to check locks: %w", err)
 					targetsCh <- nil
+					return
```
@pmazzini , @insomniacslk , @marcoguerri :
Please recheck if this return was really needed here.

tleunen commented:
# [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=h1) Report
> Merging [#60](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=desc) into [master](https://codecov.io/gh/facebookincubator/contest/commit/094f93f6a37742da445dcaf3f624679f349259cb&el=desc) will **decrease** coverage by `22.74%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/facebookincubator/contest/pull/60/graphs/tree.svg?width=650&height=150&src=pr&token=qRhZjJm6Zl)](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree)

```diff
@@             Coverage Diff             @@
##           master      #60       +/-   ##
===========================================
- Coverage   63.93%   41.19%   -22.75%     
===========================================
  Files          42       16       -26     
  Lines        2523      789     -1734     
===========================================
- Hits         1613      325     -1288     
+ Misses        689      438      -251     
+ Partials      221       26      -195     
```

| Flag | Coverage Δ | |
|---|---|---|
| #integration | `?` | |
| #unittests | `41.19% <63.25%> (+11.52%)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [pkg/target/locker.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC9sb2NrZXIuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/target/target.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC90YXJnZXQuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/pluginregistry/pluginregistry.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L3BsdWdpbnJlZ2lzdHJ5Lmdv) | `18.42% <0.00%> (-65.79%)` | :arrow_down: |
| [pkg/pluginregistry/bundles.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L2J1bmRsZXMuZ28=) | `0.00% <0.00%> (-53.97%)` | :arrow_down: |
| [pkg/lib/comparison/comparison.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2xpYi9jb21wYXJpc29uL2NvbXBhcmlzb24uZ28=) | `36.00% <0.00%> (-29.34%)` | :arrow_down: |
| [pkg/test/parameter.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3QvcGFyYW1ldGVyLmdv) | `50.00% <0.00%> (-21.43%)` | :arrow_down: |
| [pkg/test/step.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3Qvc3RlcC5nbw==) | `0.00% <0.00%> (-20.00%)` | :arrow_down: |
| [plugins/targetlocker/inmemory/inmemory.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGx1Z2lucy90YXJnZXRsb2NrZXIvaW5tZW1vcnkvaW5tZW1vcnkuZ28=) | `49.50% <0.00%> (-16.83%)` | :arrow_down: |
| [pkg/api/api.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2FwaS9hcGkuZ28=) | | |
| [pkg/job/job.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2pvYi9qb2IuZ28=) | | |
| ... and [32 more](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=footer). Last update [094f93f...73d8d9e](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

tleunen commented:
# [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=h1) Report
> Merging [#60](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=desc) into [master](https://codecov.io/gh/facebookincubator/contest/commit/094f93f6a37742da445dcaf3f624679f349259cb&el=desc) will **decrease** coverage by `22.66%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/facebookincubator/contest/pull/60/graphs/tree.svg?width=650&height=150&src=pr&token=qRhZjJm6Zl)](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree)

```diff
@@             Coverage Diff             @@
##           master      #60       +/-   ##
===========================================
- Coverage   63.93%   41.26%   -22.67%     
===========================================
  Files          42       16       -26     
  Lines        2523      790     -1733     
===========================================
- Hits         1613      326     -1287     
+ Misses        689      438      -251     
+ Partials      221       26      -195     
```

| Flag | Coverage Δ | |
|---|---|---|
| #integration | `?` | |
| #unittests | `41.26% <63.37%> (+11.60%)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [pkg/target/locker.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC9sb2NrZXIuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/target/target.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC90YXJnZXQuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/pluginregistry/pluginregistry.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L3BsdWdpbnJlZ2lzdHJ5Lmdv) | `18.42% <0.00%> (-65.79%)` | :arrow_down: |
| [pkg/pluginregistry/bundles.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L2J1bmRsZXMuZ28=) | `0.00% <0.00%> (-53.97%)` | :arrow_down: |
| [pkg/lib/comparison/comparison.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2xpYi9jb21wYXJpc29uL2NvbXBhcmlzb24uZ28=) | `36.00% <0.00%> (-29.34%)` | :arrow_down: |
| [pkg/test/parameter.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3QvcGFyYW1ldGVyLmdv) | `50.00% <0.00%> (-21.43%)` | :arrow_down: |
| [pkg/test/step.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3Qvc3RlcC5nbw==) | `0.00% <0.00%> (-20.00%)` | :arrow_down: |
| [plugins/targetlocker/inmemory/inmemory.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGx1Z2lucy90YXJnZXRsb2NrZXIvaW5tZW1vcnkvaW5tZW1vcnkuZ28=) | `49.50% <0.00%> (-16.83%)` | :arrow_down: |
| [plugins/storage/rdbms/report.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGx1Z2lucy9zdG9yYWdlL3JkYm1zL3JlcG9ydC5nbw==) | | |
| [pkg/api/api.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2FwaS9hcGkuZ28=) | | |
| ... and [32 more](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=footer). Last update [094f93f...181fb6d](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

tleunen commented:
# [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=h1) Report
> Merging [#60](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=desc) into [master](https://codecov.io/gh/facebookincubator/contest/commit/094f93f6a37742da445dcaf3f624679f349259cb&el=desc) will **decrease** coverage by `22.42%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/facebookincubator/contest/pull/60/graphs/tree.svg?width=650&height=150&src=pr&token=qRhZjJm6Zl)](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree)

```diff
@@             Coverage Diff             @@
##           master      #60       +/-   ##
===========================================
- Coverage   63.93%   41.50%   -22.43%     
===========================================
  Files          42       17       -25     
  Lines        2523      824     -1699     
===========================================
- Hits         1613      342     -1271     
+ Misses        689      456      -233     
+ Partials      221       26      -195     
```

| Flag | Coverage Δ | |
|---|---|---|
| #integration | `?` | |
| #unittests | `41.50% <63.37%> (+11.84%)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [pkg/target/locker.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC9sb2NrZXIuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/target/target.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3RhcmdldC90YXJnZXQuZ28=) | `0.00% <0.00%> (-100.00%)` | :arrow_down: |
| [pkg/pluginregistry/pluginregistry.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L3BsdWdpbnJlZ2lzdHJ5Lmdv) | `18.42% <0.00%> (-65.79%)` | :arrow_down: |
| [pkg/pluginregistry/bundles.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3BsdWdpbnJlZ2lzdHJ5L2J1bmRsZXMuZ28=) | `0.00% <0.00%> (-53.97%)` | :arrow_down: |
| [pkg/event/testevent/test.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2V2ZW50L3Rlc3RldmVudC90ZXN0Lmdv) | `47.05% <0.00%> (-41.84%)` | :arrow_down: |
| [pkg/lib/comparison/comparison.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2xpYi9jb21wYXJpc29uL2NvbXBhcmlzb24uZ28=) | `36.00% <0.00%> (-29.34%)` | :arrow_down: |
| [pkg/test/parameter.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3QvcGFyYW1ldGVyLmdv) | `50.00% <0.00%> (-21.43%)` | :arrow_down: |
| [pkg/test/step.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL3Rlc3Qvc3RlcC5nbw==) | `0.00% <0.00%> (-20.00%)` | :arrow_down: |
| [plugins/targetlocker/inmemory/inmemory.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGx1Z2lucy90YXJnZXRsb2NrZXIvaW5tZW1vcnkvaW5tZW1vcnkuZ28=) | `49.50% <0.00%> (-16.83%)` | :arrow_down: |
| [pkg/logging/logger.go](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree#diff-cGtnL2xvZ2dpbmcvbG9nZ2VyLmdv) | | |
| ... and [33 more](https://codecov.io/gh/facebookincubator/contest/pull/60/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=footer). Last update [094f93f...a52bad8](https://codecov.io/gh/facebookincubator/contest/pull/60?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

xaionaro commented:
The idea I started here with factories is complete in commit: f9008c8effd6f3279f4f45921b4c30ad663e1587

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,448 @@
+	if err == nil {
+		return
+	}
+	frame := runtimetools.Frame(1)
```
If I do so, then I will see `reportIfError` in the logs, while I was to see the function which called `reportIfError`. And I don't want to write anything to logs if `err == nil`. But yeah, I think you right, it may be an overkill to implement `runtimetools` just because I was lazy to write `if err ==nil` if few places. I should remove it.

xaionaro replied:
On the other hand. This could be handy for other pieces of the code. I will agree with any solution. I think both ways are good: as it is now (more syntax sugar) or without `runtimetools` (less code).

xaionaro replied:
Added a comment.

xaionaro replied:
**Update:**
After discussions with Andrea, I will remove this method (and it's dependency `runtimetools`).

xaionaro replied:
Removed `runtimetools`.

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,448 @@
+	tl.log.Debugf("Start transaction -> %p", tx)
+
+	defer func() {
+		if err != nil {
```
It will have the value assigned by `fn(tx)` below.

marcoguerri replied:
Well, ok, that is hard to follow. But it seems to be an idiomatic way to take actions on the error value returned by a function, so I'll accept that it's only me who is not familiar with some idioms enough.

marcoguerri replied:
I still think that trading off for readability for a reader that is not expert in all the possible Golang idioms is the preferred choice, and here I think not deferring error checking would be a tradeoff in better readability. What do you think?

xaionaro replied:
Personally I don't like to put complex logic directly to `defer` (usually it's better to declare a separate function for that and put it to the `defer`). But I think that receiving and changing the returned value through `defer` is a standard Go's feature, so it should be fine to use it sometimes.

On the other hand, it's a "plugin" and may be you right: a ConTest user (not developer) may become a potential reader of this code. OK, I'll try to avoid using this feature in here. Will be fixed.

xaionaro replied:
Refactored the function.

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,448 @@
+func (tl *TargetLocker) performActionOnTargets(
+	action targetlocker.Action,
+	jobID types.JobID,
+	targets []*target.Target,
```
I have no opinion here. I have no preference between `[]*target.Target` and `target.Targets` here. I will fix it to `target.Targets` :)

xaionaro replied:
Fixed.

xaionaro replied:
Fixed back :)

xaionaro commented on cmds/contest/main.go:
```diff
@@ -82,9 +107,37 @@ var userFunctions = map[string]interface{}{
+func expandArgument(arg string) string {
+	// it does not support correct expanding into depth more one.
+	flag.CommandLine.VisitAll(func(f *flag.Flag) {
+		arg = strings.Replace(arg, `%`+f.Name+`%`, f.Value.String(), -1)
```
> I was wondering if instead we should give the plugins a way to register flags (in separate namespaces), and then give the possibility to those plugins to consume those flags at creation time.

Yeah, indeed it would be more structured/modular/flexible. I would like to do it. You would like to see it as a part of this PR of a separate one?

xaionaro replied:
> Can you should how this would look like with multiple arguments? This seems to be fragile ...

I do not see why is this fragile [I'm not telling it's not fragile, I just miss something :) ]. It does not support of expanding of variable to a value which is also required to be expanded, but if we will meet such case we can thing this through as well.

I did-not-yet think about deep expanding because of problems like this:
```
a = `%c%`
b = `%a%`
c = `%b%`
```

marcoguerri replied:
If we go down the per plugin flags path, then definitely in a different PR.

Please, add some comment somewhere in the flag definition or here, to explain how passing arguments would look like. Maybe I missed it, but the comment for the flags says:

```
The engine to lock targets. Possible engines (the part before the first colon): %s
```

What comes after the `:`? Does it support multiple `%flag%` if separated by additional semicolons? Where are the flags that are replaced in the placeholder defined? Please help whoever is reading the code understand with as little barrier as possible. I don't think it's reasonable for me to have too look at `parseFactoryInfo`, `expandArgument`, `setupFlags` to understand how `flagTargetLocker works.

xaionaro replied:
Yeah, I right it's really confusing, thank you. I'll fix it :)

xaionaro replied:
Added more description. Please let me know if now it's clear :)

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,448 @@
+
+// TargetLocker is the no-op target locker. It does nothing.
+type TargetLocker struct {
+	log     *logrus.Entry
```
```
// GetLogger returns a configured logger instance
func GetLogger(prefix string) *logrus.Entry {
	return log.WithField("prefix", prefix)
}
```
It returns `*logrus.Entry`, so I use it here.

xaionaro commented:
> pkg/abstract: this shouldn't be necessary.

It is necessary for the next PR where I refactor factories. The problem I faced is to duplicate the logic for factories the 6th time or just to deduplicate all it. So an `abstract.Factory` appeared.

> pkg/lib/runtimetools: what is this used for?

For syntax sugar (/for code readability). With it I'm able to write:
```go
	defer func() { tl.reportIfError(validLocksRows.Close()) }()
```
Without it would have to write it:
```go
	defer func() { if err := validLocksRows.Close(); err != nil { tl.log.Error(err) } }()
```

> pkg/target/* : why do we need this new type? And since it's a type alias to a list of Target pointers, why do we need sort, len, etc ?

I use it:
```go
	allTargets.Sort()
        ...
	for validLocksRows.Next() {
                ...
		validTarget := allTargets.Find(targetID)
                ...
        }
```

I need to implement it somewhere either way (a sorted slice or a map, or something). Or I will be have to use linear search. And this way I make it reusable.

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,448 @@
+	lockTTL time.Duration
+}
+
+type backendQuerier interface {
```
Fixed.

insomniacslk commented:
> @insomniacslk :
> 
> > pkg/abstract: this shouldn't be necessary.
> 
> It is necessary for the next PR where I refactor factories. The problem I faced is to duplicate the logic for factories the 6th time or just to deduplicate all it. So an `abstract.Factory` appeared.

I've replied on #64 about this


> > pkg/lib/runtimetools: what is this used for?
> 
> For syntax sugar (/for code readability). With it I'm able to write:
> 
> ```go
> 	defer func() { tl.reportIfError(validLocksRows.Close()) }()
> ```
> 
> Without it, I would have to write it:
> 
> ```go
> 	defer func() { if err := validLocksRows.Close(); err != nil { tl.log.Error(err) } }()
> ```
> 
> I believe the sugared way is more readable/easy-to-understand (which I understand one of the main values of ConTest).

Same here, see my reply on #64 (we can continue that conversation there)


> > pkg/target/* : why do we need this new type? And since it's a type alias to a list of Target pointers, why do we need sort, len, etc ?
> 
> I use it:
> 
> ```go
> 	allTargets.Sort()
>         ...
> 	for validLocksRows.Next() {
>                 ...
> 		validTarget := allTargets.Find(targetID)
>                 ...
>         }
> ```
> 
> I need to implement it somewhere either way (a sorted slice or a map, or something). Or I will be have to use linear search. And this way I make it reusable and clean (code is not mixed-up and it's easier to read: `Target` helpers are with `Target` instead of being within a locker and other plugins).

Here we would have a `Target` type and a `Targets` type. This can become confusing quickly: which one should I use? When, and why? This can obviously confuse the users too. I am in favour of improving readability, but I'd rather have some code duplication than having two types, one of which is just syntactic sugar on top of a collection of the other type. The Go standard library tries to simplify that with things like `sort.Slice` and similar functions. I'd rather have a few top-level functions that do what you suggest, rather than a new wrapper type with glue methods.



> > without unnecessary abstractions
> 
> What is the criteria of necessity? The only global abstraction I added here is an abstract factories which called upon to eliminate 6-time code duplicate across multiple places within the code. And the code of this abstract Factory I would have to implement anyway (just in a specific factory type instead of the abstract one). So what's the point? The logic won't be reduced (except for eliminating method `ToAbstract` which seems to be undoubtedly-simple and documented enough).

We don't have a style manual, but we try to keep things as simple and readable as possible, and to always reason in that sense. To make a few examples: adding types is something we'd rather avoid. Adding an abstraction layer is something we'd avoid. Using dynamic code, unsafe code, or `interface{}` instead of a compile-time defined type is also something we'd rather avoid. There are places where we had to do that, but those are hand-picked, suffered compromises that we couldn't do otherwise (and of course there's room for improvement).

> In #64 you said you won't merge that in an assertive form (so it seems to be already decided). Just in case: please re-confirm I understood that correctly, before I will start removing `abstract.Factory` from here :)

I apologize for the tone, it's not what I intended - my reaction was "this is a huge PR with a change that doesn't seem immediately necessary, done with reflection, hey, what's going on here?". I am willing to discuss and there's no dictatorship, but of course we have to come to an agreement that serves best our users.

> **P.S.:**
> In addition to the `pkg/target/*` note:
> 
> I believe a code is much easier to read when it describes "what is required to do" instead of "how is required to do that". So instead of reimplementing sorting/finding/etc inplace (littering the code of something else) each time, I implement it only once and in a clear explicit way. So you read just `validTarget := allTargets.Find(targetID)` instead of in-place (within a plugin) code:

I find it less obvious: `allTargets.Find(targetID)` means "find target it on all targets". A better naming could be `allTargets.FindByID(targetID)`, but at that point how different is that from `findTargetByID`? Additionally, you don't have to encapsulate the simple and obvious `[]Target` object into a `Targets` object.

> ```
> func findTargetByID(targets []*Target, targetID string) *Target {
> 	idx := sort.Search(allTargets.Len(), func(i int) bool {
> 		return s[i].ID >= targetID
> 	})
> 	if idx < 0 || idx >= targets.Len() {
> 		return nil
> 	}
> 	target := targets[idx]
> 	if target.ID != targetID {
> 		return nil
> 	}
> 	return target
> }()
> 
>         ...
>         sort.Slice(len(allTargets), func(i, j) bool {
>                 return allTargets[i].ID < allTargets[j].ID
>         })
>         ...
>                  validTarget := findTargetByID(allTargets, targetID)
> ```
> 
> And you will not have to re-codereview this logic every time somebody will need it. It will be already implemented and reviewed before.

I'm tired and I think I may not be seeing what you mean - if we use `allTargets.Find(..)` vs. `findTargetsByID(...)`, where do we have to re-review the code and where's the duplicate logic?

> And it's quite easy to make a mistake when you use `sort.Search` so I wouldn't recommend to duplicate this logic. It should be implemented once, verified and then reused.

xaionaro commented:
> Same here, see my reply on #64 (we can continue that conversation there)

OK.

> Here we would have a `Target` type and a `Targets` type. This can become confusing quickly: which one should I use? When, and why?

The same as with `string` and `sort.StringSlice`.

> The Go standard library tries to simplify that with things like `sort.Slice` and similar functions.

Well, also they use `reflect` a lot. And my point here is that they have different tasks, so they have different (/appropriate) approaches. But even if it's not then there is for example `types.MethodSet` (in the standard packages) which is very similar to `target.Targets` (but they wrapped the slice into a `struct` and they perform more difficult stuff). Or for example `ed25519.PrivateKey` (see method `Seed()`), or `unicode.SpecialCase`, etc. Each of them just a new type which implements handy stuff, so you may just cast it and use it. Although, it's very different from our case, but the task is very different as well.

In total I believe this argument is arguable.

> Additionally, you don't have to encapsulate the simple and obvious []Target object into a Targets object.

In Go inter-package communication mostly is based on interfaces. So everything is represented as a type which can do something. Functional approach (you propose here) started to appear in Go recently and as [an additional syntax sugar](https://github.com/golang/go/issues/16721) (to do not reimplement `Less` and stuff all the time) if I remember correctly. And there _was not other standard way_ to sort a slice but to declare a type for the slice (may be this understanding is already considered obsolete, but I've never heard about it).

Also you may see that `reflect` for example does not have `FuncByName` (it has only `MethodByName`). And any method is easy to override (which is required for late-binding conception of OOP), while you cannot override a function. And standard packages loves to accept interfaces (not functions, usually). Even `error` is an interface. My point is declaring a type and methods is a more Golang-y and more flexible (in Go) way to handle stuff in my very humble opinion. But of course if it is possible to implement something using basic types (builtin types) then it's preferred ("require less, provide more"). But if you already use a slice of a custom type (`[]*target.Target`) then why not to declare methods right to it? Because of additional `type Targets []*Target`? I don't see it as a strong enough reason to ignore all that flexibility and sugar (while good sugar improves readability) we receive due to this one line. Also I don't see it as a strong enough reason to just reduce amount of lines. The thing that should be simplified is logic, not amount of lines. And there's nothing bad to write a simple logic in multiple lines (with extra error types, extra helper types, extra comments, extra unit tests and stuff), and as an example I believe that "hardly-readable one-liner" (less lines with the same logic) are evil.

Also I just made an experiment, I opened the alphabetically first source code file of Go (it was `archive/tar/common.go`). And it starts with `type headerError []string` :)... and then there stuff returns interfaces and accept interfaces (so abstract communication is handled through communicating with subset of methods instead of communicating with functions).

I can continue this further, but I believe this is already too much for one iteration of this discussion :)

Although, another story is if we want to write ConTest for people who a not gophers (for people who are not familiar which Go tricks and customs). If so then just let's say so :)

As another approach, we may implement functions within `lib/`:
```go
func SortByString(length int, valueFunc func(idx int) string) {
}

func SearchByString(length int, searchForValue string, valueFunc func(idx int) string) int {
	return sort.Search(s.Len(), func(i int) bool {
		return valueFunc() >= targetID
	})
}

func FindByString(length int, searchForValue string, valueFunc func(idx int) string) int {
	idx := SearchByString(length, searchForValue, valueFunc)
	if idx < 0 || idx >= s.Len() {
		return -1
	}
	target := s[idx]
	if valueFunc(idx) != searchForValue {
		return -1
	}
	return idx
}
```
```go
func someFunc() {
    ...
    idx := slicetools.FindByString(len(targets), targetID, func(idx int) string { return targets[idx].ID } )
    if idx <0 {
        return notFound
    }
    target := targets[idx]
    ...
}
```
This is the approach of communicating with functions (like `sort.Slice`), but I wouldn't say it's more readable than the current one:
```go
type Targets []*Target

func (s Targets) Search(targetID string) int {
    return sort.Search(s.Len(), func(i int) bool {
        return s[i].ID >= targetID
    })
}

func (s Targets) Find(targetID string) *Target {
    idx := s.Search(targetID)
    if idx < 0 || idx >= s.Len() {
        return nil
    }
    target := s[idx]
    if target.ID != targetID {
        return nil
    }
    return target
}
```
```go
func someFunc() {
    ...
    target := targets.Find(targetID)
    if target == nil {
        return notFound
    }
    ...
}
```

Another approach:
```go
func Search(s []*target.Target, targetID string) int {
    return sort.Search(s.Len(), func(i int) bool {
        return s[i].ID >= targetID
    })
}

func Find(s []*target.Target, targetID) {
    idx := SearchByID(s, targetID)
    if idx < 0 || idx >= s.Len() {
        return nil
    }
    target := s[idx]
    if target.ID != targetID {
        return nil
    }
    return target
}
```
```go
func someFunc() {
    ...
    target := pkgtarget.Find(targets, targetID)
    if target == nil {
        return notFound
    }
    ...
}
```
(but it is effectively the same as in PR)

Or another approach is just duplicate the logic and write it in the inline manner. And as I said it's easy to make a mistake when you use `sort.Search`, so it would be nice to implement it once, check and then reuse.

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,527 @@
+}
+
+// do performs a Lock, Unlock or RefreshLock (depending on value of `action`).
+func (q mysqlLocksQuerier) do(
```
This function is no longer required (after recent changes). I will remove it.

xaionaro commented on plugins/targetlocker/mysql/mysql_target_locker.go:
```diff
@@ -0,0 +1,527 @@
+	if err != nil {
+		return nil, targetlocker.ErrInvalidDSN{Err: err}
+	}
+	mysqlCfg.MultiStatements = true
```
Whoops. It's not required anymore. Forgot to remove it.

xaionaro replied:
Removed.

insomniacslk commented:
At this point I think we just have diverging opinions. For the sake of progressing, I'm OK with the Targets type here

xaionaro commented:
> At this point I think we just have diverging opinions. For the sake of progressing, I'm OK with the Targets type here

If I understand correctly my arguments hasn't convinced you. I'll fix it to be inline.

xaionaro commented on pkg/target/targets.go:
```diff
@@ -0,0 +1,61 @@
+
+// Targets is a helper to manipulate with a slice of Targets
+// (it collects common routines for multiple Targets).
+type Targets []*Target
```
I don't want to continue wasting time on discussion on this PR, so to comply review requirements I removed this helpers and just added an in-place implementation within `plugins/targetlocker/mysql/mysql_target_locker.go`...

xaionaro replied:
Removed type `Targets`.

xaionaro commented:
> pkg/abstract: this shouldn't be necessary. As much as I like abstractions, this seems unnecessary and only adding complexity. The factory interface can be replaced by a loader function (similar to what we do for other plugin types). Code duplication is sometimes a better compromise

Removed `pkg/abstract`.

> pkg/lib/runtimetools: what is this used for?

Removed `pkg/lib/runtimetools`

> pkg/target/* : why do we need this new type? And since it's a type alias to a list of Target pointers, why do we need sort, len, etc ? These could make sense if the slice was private, but it's not the case here (and there's no point for making it private), so this is just unnecessary boilerplate

Removed type `target.Targets`.

xaionaro commented:
Should we close it and start from scratch or you would prefer me to fix here something? I'm open to any suggestions. I just want to either merge it or close it. So just say what to do :)

tfg13 commented:
Done in a simplified version in #136
